### PR TITLE
Fix fixable `ruff` rules

### PR DESCRIPTION
**Pull request by DanielYang59**
### Summary

Fix fixable `ruff` rules

### TODOs
- [x]  Unit test of `io.aims` doesn't follow name convention, and has test files inside it.
- [x] Fix NPY201 https://github.com/materialsproject/pymatgen/blob/41e4c69911f2b724734e0917c436ee5a1de63091/pyproject.toml#L209
- [x] Fix PD011 (there might be a few false-positive for example in `io.abinit.pseudo` and io.cp2k keywords, https://github.com/astral-sh/ruff/issues/6432)
- [x] [**Need help**] Use of relative import in test cf24f12ccd5adf5c34a9c32899f87b2f790effe0

### Future PR
- Replace `http` URLs with `https`

### CODE REVIEW

File: tests/io/aims/conftest.py
Comment (DanielYang59) [inline tests/io/aims/conftest.py:27]: Need confirm: these helper functions for io.aims test is now relocated from `pymatgen.util.testing.aims`, is this intended? @shyuep

File: tests/io/aims/sets/test_bs_generator.py
Comment (DanielYang59) [inline tests/io/aims/sets/test_bs_generator.py:8]: Not sure how to import without using relative import in this case...

File: src/pymatgen/io/vasp/help.py
Comment (DanielYang59) [inline src/pymatgen/io/vasp/help.py:58]: Need confirm, I can see this helper function was added with SSL certificate turned off (#1791), not sure if this is intended for any reason. The `print_help` method seems to work fine with `verify` enabled:
```python
from pymatgen.io.vasp.help import VaspDoc


doc = VaspDoc()
doc.print_help("ISYM")
```

File: src/pymatgen/io/vasp/help.py
Comment (DanielYang59) [inline src/pymatgen/io/vasp/help.py:77]: The `get_incar_tags` method doesn't seem to work with or without `verify`:
```
developer/pymatgen/src/pymatgen/io/vasp/help.py:77: GuessedAtParserWarning: No parser was explicitly specified, so I'm using the best available HTML parser for this system ("html.parser"). This usually isn't a problem, but if you run this code on another system, or in a different virtual environment, it may use a different parser and behave differently.

The code that caused this warning is on line 77 of the file /Users/yang/developer/pymatgen/src/pymatgen/io/vasp/help.py. To get rid of this warning, pass the additional argument 'features="html.parser"' to the BeautifulSoup constructor.

  soup = BeautifulSoup(response.text)
```

File: src/pymatgen/io/abinit/abiobjects.py
Comment (janosh) [inline src/pymatgen/io/abinit/abiobjects.py:341]: maybe call this `SpinModeTuple` or `SpinModeNT`

File: src/pymatgen/analysis/adsorption.py
Comment (janosh) [inline src/pymatgen/analysis/adsorption.py:305]: generally not so sure this ruff rule is actually an improvement. more verbose and less readable imo. maybe just ignore `RUF017` in general?

File: tests/io/aims/sets/test_static_restart_from_relax_generator.py
Comment (janosh) [inline tests/io/aims/sets/test_static_restart_from_relax_generator.py:8]:
```suggestion
from tests.conftest import O2, Si, comp_system
```

**Review by janosh — CHANGES_REQUESTED** [review]
thanks, good stuff as always!

will hold off on reviewing #4017 until this is merged

File: src/pymatgen/io/vasp/help.py
Comment (janosh) [reply src/pymatgen/io/vasp/help.py]: academic websites sometimes lack SSL certificates. perhaps VASP docs did in the past, but since they have SSL now, no reason not to enable

File: src/pymatgen/io/abinit/abiobjects.py
Comment (DanielYang59) [reply src/pymatgen/io/abinit/abiobjects.py:341]: ok!

File: tests/io/aims/sets/test_static_restart_from_relax_generator.py
Comment (DanielYang59) [reply tests/io/aims/sets/test_static_restart_from_relax_generator.py:8]: Can you double check this? This doesn't seem to work (we have to import from `pymatgen/tests/io/aims/conftest.py` not `pymatgen/tests/conftest.py`). Also `pymatgen.tests` isn't made a module at this moment (not sure if absolute import would work).

```python
____________________________________ ERROR collecting tests/io/aims/sets/test_bs_generator.py ____________________________________
ImportError while importing test module '/Users/yang/developer/pymatgen/tests/io/aims/sets/test_bs_generator.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
tests/io/aims/sets/test_bs_generator.py:8: in <module>
    from tests.conftest import Si, comp_system  # noqa: TID252
E   ImportError: cannot import name 'Si' from 'tests.conftest' (/Users/yang/developer/pymatgen/tests/conftest.py)
```

File: src/pymatgen/analysis/adsorption.py
Comment (DanielYang59) [reply src/pymatgen/analysis/adsorption.py:305]: I think this ruff is implemented to reduce time complexity ([`Checks for the use of sum() to flatten lists of lists, which has quadratic complexity.`](https://docs.astral.sh/ruff/rules/quadratic-list-summation/)), happy to revert this if you believe readability is more important than performance for these cases.

File: tests/io/aims/sets/test_static_restart_from_relax_generator.py
Comment (janosh) [reply tests/io/aims/sets/test_static_restart_from_relax_generator.py:8]: ah, then it should work with `from tests.io.aims.conftest import ...`

File: src/pymatgen/analysis/adsorption.py
Comment (janosh) [reply src/pymatgen/analysis/adsorption.py:305]: good to know. didn't realize `sum` scales differently. i would say let's go with `[item for sublist in lists for item in sublist]` then since it doesn't require extra imports and list comprehension is often the most performant in python

File: tests/io/aims/sets/test_static_restart_from_relax_generator.py
Comment (DanielYang59) [reply tests/io/aims/sets/test_static_restart_from_relax_generator.py:8]: It doesn't seem to work on my machine either. I thought we could only:
1. Relative import, which is used right now but the `ruff` rule TID252 would suggest against this.
2. Absolute import, with `pymatgen.tests.io.aims.conftest`, but this would not work because `pymatgen.tests` are not a package in our case (no `__init__.py`).

Perhaps just ignore this rule for tests?
> However, explicit relative imports are an acceptable alternative to absolute imports, especially when dealing with complex package layouts where using absolute imports would be unnecessarily verbose:

Comment (janosh) [reply tests/io/aims/sets/test_static_restart_from_relax_generator.py:8]: > Perhaps just ignore this rule for tests?

sounds good to me. the absolute import from[ `tests` works in `pymatviz`](https://github.com/janosh/pymatviz/blob/1de350a146b53ae03f343ff9a78454ba7c976186/tests/test_bar.py#L12) but that's because it has an `__init__.py`

File: src/pymatgen/analysis/adsorption.py
Comment (DanielYang59) [reply src/pymatgen/analysis/adsorption.py:305]: Wonderful idea. It looks like [list comprehension performs pretty well](https://realpython.com/python-flatten-list/#considering-performance-while-flattening-your-lists):

```
Time to flatten a 1000x1000 matrix (in milliseconds):

flatten_concatenation()..........1.95 ms
flatten_extend().................2.03 ms
flatten_reduce_iconcat().........2.68 ms
flatten_chain()..................4.60 ms
flatten_comprehension()..........7.79 ms  >> with list comprehension 
flatten_sum().................1113.22 ms
flatten_reduce_concat().......1117.15 ms
flatten_reduce_lambda().......1117.52 ms
flatten_reduce_add()..........1118.80 ms
```

File: tests/io/aims/sets/test_static_restart_from_relax_generator.py
Comment (DanielYang59) [reply tests/io/aims/sets/test_static_restart_from_relax_generator.py:8]: ok!

**Review by janosh — APPROVED** [review]
thanks a million @DanielYang59! 👍

**Comment by DanielYang59** [comment]
No problem, thanks for reviewing!

**Pull request merged**
